android/kotlin/v1/AmpliApp/app/src/main/java/com/amplitude/ampli/Ampli.kt

```diff
@@ -519,6 +518,30 @@ open class Ampli {
         }
 
         this._client?.setPlan(options?.client?.plan ?: observePlan)
+
+        // set IngestionMetadata with backwards compatibility, min Android SDK version 2.38.2.
+        try {
+            val clazz = Class.forName("com.amplitude.api.IngestionMetadata")
```

Comment:
@liuyang1520 @yuhao900914 @qingzhuozhen 
In general I would like us to make codegen as dumb and decoupled as possible. Anything we can move to the static SDKs (that can be unit tested, etc) is preferable to adding to codegen. I release we did similar type changes in python, js/ts but this implementation feels more tightly coupled and could lead to maintenance issues long term.

For these reasons I think the plugin/extra approach is preferable. Happy to discuss more.

Reply:
Thanks for reviewing! I reply here for @qingzhuozhen 's question too.

This solution includes the Kotlin Android SDK v1/v2 change. We tried the reflection on Kotlin Android v2 at first, and thought it doesn't work. After we figured out the reflection Java SDK, we found out the solution for Kotlin yesterday with the inspiration of Java SDK.

I prefer the reflection way, due to the following reasons:
- No other releases needed for SDKs, also legacy SDKs are using middleware, which need more efforts to explore where/how to attach the information
- Clean implementation, doesn't need to manipulate the `extra` field (I thought that was the only solution before), I think technically this field should be used by customer for adding information, this way we can avoid using it
- Low maintenance efforts, if anything goes wrong with ingestion metadata, this would be the single place to check in Ampli codegen. In the future, if we assure customers are using the latest version SDKs, we can safely replace this with the direct assigning of ingestion metadata, no need to update SDK anymore

The ingestion metadata functionality has been unit tested in static SDKs, this is to use it with supporting the backwards compatibility. In the long term, think we can explore a better way to release the codegen change (another story).

Sorry I haven't spent many efforts to investigate the reflection in the beginning, I think I should have tried that more before drawing the conclusion.

Happy to discuss/address more concerns! Thank you for reviewing!